projects/plugins/jetpack/modules/shortcodes/smartframe.php

```diff
@@ -0,0 +1,124 @@
+<?php
+/**
+ * Smartframe.io embed
+ *
+ * Example URL: https://mikael-korpela.smartframe.io/p/mantymetsa_1630927773870/7673dc41a775fb845cc26acf24f1fe4?t=rql1c6dbpv2
+ * Example embed code: <script src="https://embed.smartframe.io/6ae67829d1264ee0ea6071a788940eae.js" data-image-id="mantymetsa_1630927773870" data-width="100%" data-max-width="1412px"></script>
+ *
+ * @package automattic/jetpack
+ */
+
+if ( defined( 'IS_WPCOM' ) && IS_WPCOM ) {
+	add_action( 'init', 'jetpack_smartframe_enable_embeds' );
+} else {
+	jetpack_smartframe_enable_embeds();
+}
+
+/**
+ * Register smartframe as oembed provider. Add filter to reverse iframes to shortcode. Register [smartframe] shortcode.
+ *
+ * @since 9.3.3
+ */
+function jetpack_smartframe_enable_embeds() {
+	// Support their oEmbed Endpoint.
+	wp_oembed_add_provider( '#https?://(.*?)\.smartframe\.(io|net)/.*#i', 'https://oembed.smartframe.io/', true );
+
+	// Allow script to be filtered to short code (so direct copy+paste can be done).
+	add_filter( 'the_content', 'wpcom_shortcodereverse_smartframe' );
+
+	// Actually display the smartframe Embed.
+	add_shortcode( 'smartframe', 'jetpack_smartframe_shortcode' );
+}
+
+/**
+ * Compose shortcode based on smartframe iframes.
+ *
+ * @since 9.3.3
+ *
+ * @param string $content Post content.
+ *
+ * @return mixed
+ */
+function wpcom_shortcodereverse_smartframe( $content ) {
+	if ( ! is_string( $content ) || false === stripos( $content, 'embed.smartframe' ) ) {
+		return $content;
+	}
+
+	// phpcs:ignore WordPress.WP.EnqueuedResources.NonEnqueuedScript
+	$regexp     = '!<script\ssrc="https://embed\.smartframe\.(io|net)/(\w+)\.js"\sdata-image-id="(.*?)"(\sdata-width="(\d+(%|px))"\s)?(data-max-width="(\d+(%|px)))?"></script>!i';
+	$regexp_ent = str_replace( '&amp;#0*58;', '&amp;#0*58;|&#0*58;', htmlspecialchars( $regexp, ENT_NOQUOTES ) );
+
+	foreach ( compact( 'regexp', 'regexp_ent' ) as $regexp ) {
+		if ( ! preg_match_all( $regexp, $content, $matches, PREG_SET_ORDER ) ) {
+			continue;
+		}
+
+		foreach ( $matches as $match ) {
+			$script_id = $match[2];
+
+			if ( empty( $script_id ) ) {
+				continue;
+			}
+
+			$image_id = $match[3];
+
+			if ( empty( $image_id ) ) {
+				continue;
+			}
+
+			if ( isset( $match[5] ) ) {
+				$width = $match[5];
+			}
+
+			if ( isset( $match[8] ) ) {
+				$max_width = $match[8];
+			}
+
+			$shortcode = '[smartframe script-id="' . $script_id . '" image-id="' . esc_attr( $image_id ) . '"';
+			if ( ! empty( $width ) ) {
+				$shortcode .= ' width="' . esc_attr( $width ) . '"';
+			}
+			if ( ! empty( $max_width ) ) {
+				$shortcode .= ' max-width="' . esc_attr( $max_width ) . '"';
+			}
+			$shortcode .= ']';
```

Comment:
Part of me wonders if it wouldn't be simpler to do something like this?

```suggestion
			// We need at least a script and an image ID.
			if ( ! isset( $match[2], $match[3] ) ) {
				continue;
			}

			$shortcode = sprintf(
				'[smartframe script-id="%1$s" image-id="%2$s"%3$s]',
				esc_attr( $match[2] ),
				esc_attr( $match[3] ),
				! empty( $match[8] ) ? ' max-width="' . esc_attr( $match[8] ) . '"' : ''
			);
```

It also has the benefit of clearly showing that everything is escaped.

(this suggestion also removes `width`, since we do not appear to be using it at the moment?)